mempool/cat/cache.go

```diff
@@ -0,0 +1,292 @@
+		timeEvicted: time.Now().UTC(),
+		priority:    wtx.priority,
+		gasWanted:   wtx.gasWanted,
+		sender:      wtx.sender,
+		size:        wtx.size(),
+	}
+	// if cache too large, remove the oldest entry
+	if len(c.cache) > c.size {
+		oldestTxKey := wtx.key
+		oldestTxTime := time.Now().UTC()
+		for key, info := range c.cache {
+			if info.timeEvicted.Before(oldestTxTime) {
+				oldestTxTime = info.timeEvicted
+				oldestTxKey = key
+			}
+		}
+		delete(c.cache, oldestTxKey)
+	}
+}
+
+func (c *EvictedTxCache) Pop(txKey types.TxKey) *EvictedTxInfo {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	info, exists := c.cache[txKey]
+	if !exists {
+		return nil
+	} else {
+		delete(c.cache, txKey)
+		return info
+	}
+}
+
+func (c *EvictedTxCache) Prune(limit time.Time) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	for key, info := range c.cache {
+		if info.timeEvicted.Before(limit) {
+			delete(c.cache, key)
+		}
+	}
+}
+
+func (c *EvictedTxCache) Reset() {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	c.cache = make(map[types.TxKey]*EvictedTxInfo)
+}
+
+// SeenTxSet records transactions that have been
+// seen by other peers but not yet by us
+type SeenTxSet struct {
+	mtx tmsync.Mutex
+	set map[types.TxKey]timestampedPeerSet
+}
+
+type timestampedPeerSet struct {
+	peers map[uint16]bool
+	time  time.Time
+}
+
+func NewSeenTxSet() *SeenTxSet {
+	return &SeenTxSet{
+		set: make(map[types.TxKey]timestampedPeerSet),
+	}
+}
+
+func (s *SeenTxSet) Add(txKey types.TxKey, peer uint16) {
+	if peer == 0 {
+		return
+	}
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	seenSet, exists := s.set[txKey]
+	if !exists {
+		s.set[txKey] = timestampedPeerSet{
+			peers: map[uint16]bool{peer: true},
+			time:  time.Now().UTC(),
+		}
+	} else {
+		seenSet.peers[peer] = true
+	}
+}
+
+func (s *SeenTxSet) Pop(txKey types.TxKey) uint16 {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	seenSet, exists := s.set[txKey]
+	if !exists {
+		return 0
+	} else {
+		for peer := range seenSet.peers {
+			delete(seenSet.peers, peer)
+			return peer
+		}
+		return 0
+	}
+}
+
+func (s *SeenTxSet) RemoveKey(txKey types.TxKey) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	delete(s.set, txKey)
+}
+
+func (s *SeenTxSet) Remove(txKey types.TxKey, peer uint16) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	set, exists := s.set[txKey]
+	if exists {
+		if len(set.peers) == 1 {
+			delete(s.set, txKey)
+		} else {
+			delete(set.peers, peer)
+		}
+	}
+}
+
+func (s *SeenTxSet) Prune(limit time.Time) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	for key, seenSet := range s.set {
+		if seenSet.time.Before(limit) {
+			delete(s.set, key)
+		}
+	}
+}
+
+func (s *SeenTxSet) Has(txKey types.TxKey, peer uint16) bool {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	seenSet, exists := s.set[txKey]
+	if !exists {
+		return false
+	} else {
+		return seenSet.peers[peer]
+	}
+}
+
+func (s *SeenTxSet) Get(txKey types.TxKey) map[uint16]struct{} {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	seenSet, exists := s.set[txKey]
+	if !exists {
```

Comment:
same, if you reverse this you can remove some indentation to make it more inline and readable imo